src/Build/Evaluation/Expander.cs

```diff
@@ -1318,51 +1334,53 @@ internal static string ConvertToString(object valueToConvert)
                     {
                         convertedString = (string)valueToConvert;
                     }
-                    else if (valueToConvert is IDictionary)
+                    else if (valueToConvert is IDictionary dictionary)
                     {
                         // If the return type is an IDictionary, then we convert this to
                         // a semi-colon delimited set of A=B pairs.
                         // Key and Value are converted to string and escaped
-                        IDictionary dictionary = valueToConvert as IDictionary;
-                        using (var builder = new ReuseableStringBuilder())
+                        if (dictionary.Count > 0)
                         {
+                            using SpanBasedStringBuilder builder = Strings.GetSpanBasedStringBuilder();
+
                             foreach (DictionaryEntry entry in dictionary)
                             {
                                 if (builder.Length > 0)
                                 {
-                                    builder.Append(';');
```

Comment:
In this first iteration I'd like to keep it simple and support only real spans. Making a span out of a character is problematic. `ch.ToString().AsSpan()` allocates so that's a no-go. Maybe the caller could pass the character by-ref so it could be converted to span without allocating, or the `SpanBasedStringBuilder` could keep a list of chars in addition to the list of spans but it feels like it would start defeating the purpose. Good question, though.